src/commands/force/source/beta/pull.ts

```diff
@@ -0,0 +1,185 @@
+ */
+
+import { FlagsConfig, flags } from '@salesforce/command';
+import { Duration } from '@salesforce/kit';
+import { Messages } from '@salesforce/core';
+import {
+  FileResponse,
+  SourceComponent,
+  ComponentSet,
+  RetrieveResult,
+  RequestStatus,
+  ComponentStatus,
+} from '@salesforce/source-deploy-retrieve';
+import { SourceTracking, throwIfInvalid, replaceRenamedCommands, ChangeResult } from '@salesforce/source-tracking';
+import { processConflicts } from '../../../../formatters/conflicts';
+import { SourceCommand } from '../../../../sourceCommand';
+import { PullResponse, PullResultFormatter } from '../../../../formatters/pullFormatter';
+Messages.importMessagesDirectory(__dirname);
+const messages: Messages = Messages.loadMessages('@salesforce/plugin-source', 'pull');
+
+export default class SourcePull extends SourceCommand {
+  public static description = messages.getMessage('description');
+  public static help = messages.getMessage('help');
+  protected static readonly flagsConfig: FlagsConfig = {
+    forceoverwrite: flags.boolean({
+      char: 'f',
+      description: messages.getMessage('flags.forceoverwrite'),
+    }),
+    // TODO: use shared flags from plugin-source
+    wait: flags.minutes({
+      char: 'w',
+      default: Duration.minutes(33),
+      min: Duration.minutes(0), // wait=0 means deploy is asynchronous
+      description: messages.getMessage('flags.waitLong'),
+    }),
+  };
+
+  protected static requiresUsername = true;
+  protected static requiresProject = true;
+  protected hidden = true;
+  protected readonly lifecycleEventNames = ['preretrieve', 'postretrieve'];
+  protected tracking: SourceTracking;
+  protected retrieveResult: RetrieveResult;
+  protected deleteFileResponses: FileResponse[];
+
+  public async run(): Promise<PullResponse[]> {
+    await this.preChecks();
+    await this.retrieve();
+    // do not parallelize delete and retrieve...we only get to delete IF retrieve was successful
+    await this.doDeletes(); // deletes includes its tracking file operations
+    await this.updateTrackingFilesWithRetrieve();
+    this.ux.stopSpinner();
+
+    return this.formatResult();
+  }
+
+  protected async preChecks(): Promise<void> {
+    // checks the source tracking file version and throws if they're toolbelt's old version
+    throwIfInvalid({
+      org: this.org,
+      projectPath: this.project.getPath(),
+      toValidate: 'plugin-source',
+      command: replaceRenamedCommands('force:source:pull'),
+    });
+
+    this.ux.startSpinner('Loading source tracking information');
+    this.tracking = await SourceTracking.create({
+      org: this.org,
+      project: this.project,
+    });
+
+    await this.tracking.ensureRemoteTracking(true);
+    this.ux.setSpinnerStatus('Checking for conflicts');
+
+    if (!this.flags.forceoverwrite) {
+      processConflicts(await this.tracking.getConflicts(), this.ux, messages.getMessage('sourceConflictDetected'));
```

Comment:
I don't love passing `this.ux` to STL.  I would expect STL to just return the conflicts and the command would handle it how it wants.  If it's this way so that code can be shared and the sf plugin won't need to also implement it, what will happen if sf (or any other plugin) wants to process conflicts differently, or even have a different message?

Reply:
`processConflicts` is in the plugin's /formatters, not in STL.  It's passing output from `STL.getConflicts()` to the processConflicts "formatter" (and I use that term pretty loosely)

Reply:
Ahh, ok.  I missed that.